### PR TITLE
feat: Add multiple table filters for Row Level Security

### PR DESCRIPTION
**Pull request by axelet**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [x] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY

Introduces multiple table row level security filters instead of one table filters.

In a nutshell introduces a new table called **rls_filter_tables**. It represents the relation between rls_filter_id and table_id pretty much the same as we have for role-filter relations in the **rls_filter_roles** table.

### BEFORE SCREENSHOTS

Screenshot 2020-05-06 at 18 28 34
Screenshot 2020-05-06 at 18 28 46

### AFTER SCREENSHOTS
Screenshot 2020-05-06 at 17 45 44
Screenshot 2020-05-06 at 17 45 13

### ADDITIONAL INFORMATION

- [x] Has associated issue: #9532 
- [ ] Changes UI
- [x] Requires DB Migration.
- [x] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: docker/pythonpath_dev/superset_config.py [hunk 1]
```diff
@@ -67,6 +67,7 @@ def get_env_variable(var_name, default=None):
 
 RESULTS_BACKEND = FileSystemCache('/app/superset_home/sqllab')
 
+ENABLE_ROW_LEVEL_SECURITY = True
```
Comment (villebro) [inline docker/pythonpath_dev/superset_config.py:70]: This should not be enabled by default.

**Review by villebro — COMMENTED** [review]
Quick first pass. Would it be possible to add a few screenshots showing what the user experience for adding multi-table RLS looks like?

File: docker/pythonpath_dev/superset_config.py (hunk 1 above)
Comment (axelet) [reply docker/pythonpath_dev/superset_config.py:70]: Sure

**Comment by axelet** [comment]
@villebro Is it necessary to migrate possible **row_level_security_filters** entries since the RLS feature is not considered prod quality for now? Do we have to support the migration anyway? Because now the update only makes db structure migrations and upgrading/downgrading with entries can cause some issues. It only support migration with empty row_level_security_filters.

**Comment by villebro** [comment]
@axelet As we know this feature is already in production use in some places, I think we should provide proper migrations to support non-breaking upgrades.

**Comment by axelet** [comment]
@villebro Hey, I've made 2 commits and both of them are suitable for records migration. 

The difference is that e73e3ce support only old-fashioned records downgrading ("old-fashioned" - records having 1 table_id per rls_filter_id). The last commit https://github.com/apache/incubator-superset/pull/9751/commits/88d7206a9a059f488eb6eed0723d6e69bf55bc93 supports downgrading new-fashioned records. There is a case when we can have multiple table_ids corresponding to 1 rls_filter_id. So, to downgrade this kind of records we have to add as much filter copies to row_level_security_filters as the rls_filter_tables have records related to the filter. Then we have to make sure roles are pointing to the newly created filters respectively and create new role-filter records in rls_filter_roles if needed.

For example, given the screenshot:
Screenshot 2020-06-03 at 16 34 41
We have 3 flattened (old-fashioned) records and 1 new-fashioned that we have to flatten.
The new-fashoined record have 2 entries in rls_filter_tables, 1 entry in row_level_security_filters, and 2 entry in rls_filter_roles. To return to old tables structure we lack 1 filter entry row_level_security to be able to populate it by rls_filter_tables's table_id. As we create it and assign a table_id we now have to assign similar roles to the filter. That will add 2 more records to rls_filter_roles to correspond the newly created filter. As for those records that have 1-to-1 table_id to filter_id relations we just move these table_ids to row_level_security_filter table respectively. And there is no need to do anything with roles.

And the result of downgrading will look like this:
Screenshot 2020-06-03 at 16 43 35

File: superset/migrations/versions/be1e5bcb9f46_.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import sqlalchemy as sa
+from alembic import op
+
+
+def upgrade():
+    pass
+
+
+def downgrade():
+    pass
```
Comment (villebro) [inline superset/migrations/versions/be1e5bcb9f46_.py:38]: I think this migration is redundant, and should be avoidable by pointing `e557699a813e` to `f9a30386bd74`

**Review by villebro — CHANGES_REQUESTED** [review]
Thanks for making the prudent downgrade migration! I suggest removing the redundant migration. Apart from that this LGTM, I can test this in the coming days.

File: superset/migrations/versions/be1e5bcb9f46_.py (hunk 2 above)
Comment (axelet) [reply superset/migrations/versions/be1e5bcb9f46_.py:38]: Yeap, sure, that's a good point, will do

**Comment by villebro** [comment]
Flagging this for the upcoming 0.37 release. Sorry for the delay, I will review and test this now.

**Review by villebro — CHANGES_REQUESTED** [review]
This works beautifully! Please fix the `down_revision` id, but apart from that this seems good to go.

**Review by villebro — APPROVED** [review]
Thanks for your patience and fixing the migration id; LGTM, merging now.

**Comment by axelet** [comment]
Thank you for your time discussing and testing it!

**Pull request merged**